### PR TITLE
Improvements to the Enter Dialog

### PR DESCRIPTION
**Pull request by Wycklife**
Fixes the last two parts of #5596

1. Swapping the calculator basic keyboard and the enter basic keyboard
2. Added keys and improvements to the existing enter keyboard
3. Added examples and tool tips to the list view linked  with the example button.
4. Replaced the old save control with the new control and set the positioning of the variables

@africanmathsinitiative/developers This is ready for review

### CODE REVIEW

**Comment by Wycklife** [comment]
@lilyclements This is ready for you to look at the design first.  Thanks.

**Comment by lilyclements** [comment]
@Wycklife can you update this branch with the master? Then I can make some design changes and push them into your branch (if you're happy with them)

**Comment by Wycklife** [comment]
@lilyclements Okay I am doing that.

**Comment by Wycklife** [comment]
@lilyclements The branch has been updated

**Comment by rdstern** [comment]
@Wycklife this has made good progress.  I mention what has been done well and also what remains - I should say that some of what remains is because of errors I made in the instructions.

a) The new examples have been added together with the tooltips.  Please now detete the original examples, they are the top 4 ones, down to month.abb.
b) The new save control with the position added, is included.  Great.
c) There is an oddity to try to fix.  The first examples now are 4.5 and 'Nairobi'.  @Wycklife I wonder if you tested them?  If so, you might have noticed that the Try control doesn't work.  It says they are an error - but they are not, in the sense that they do give an acceptable variable.  Can you fix the try control?  This is something that @Patowhiz has examined for another dialogue, so he may be able to help if you get stuck.
d) The main changes are to the second keyboard.  I need to add tooltips and there should also be changes in the keyboard.
1) the keyboard should move down a bit, so the keys are alighted with the main keyboard.
2) The keys should become narrower, so the dialogue, as a whole is not so wide.
3) Sorry, I do need to change some of the keys and also the layout.  Your putting them into alphabetical order is sensible, but I propose below a new set and order.
4. It remains as 16 keys and I list them in the next message going across:

**Comment by rdstern** [comment]
Row 1:
a) Change : to `[ :]` with the cursor before the colon.  Tooltip A subset, e.g. letters[1:4] gives a, b, c, d.
b) The second is `letters` as now. Tooltip: The letters, e.g. letters[1:4] gives a, b, c, d. Type LETTERS[1:4] for A, B, C, D.
c) factor, to now give `forcats::as_factor( )`. Tooltip: e.g. forcats::as_factor("B","C","A") to make the contents into a factor variable with the levels in the order as entered, so here B is the lowest level.

(This reminds me.  Can you change the second factor example to forcats::as_factor(rep(LETTERS[4:1],c(4,3,1,1))) with tooltip changed to Gives D,D,D,D,C,C,C,B,A as a fa ctor.)
d) rnorm as before, giving rnorm( ) as now, but also adding the parameters when that is ticked. Tooltip: Random sample from a normal distribution, e.g. rnorm(36, 100, 15) gives 36 observations from a distribution with mean 100 and s.d. 15; rnorm(36) is from a standard normal distribution.
More to come.

**Comment by Wycklife** [comment]
@rdstern I have fixed items 1 to 4 above. On the try conrol, I have gone through the control and realized that the control only accept commands not single elements. I'm not sure on how to add this an additional property to the control for it to accept these elements. Though when I convert them to vectors i.e c(4.5) and c('Nairobi') the control seems to work. Should I pass these items as vectors?

**Comment by Wycklife** [comment]
image

This is the new layout of the keyboard and the dialog.

**Comment by rdstern** [comment]
@Wycklife this layout looks neat and I look forward to testing it.
On your solution for the Try with a single element, I see it works on the command, perhaps because the number, or string is put into an object and then used.  
In a similar situation (File > New Data Frame) @Patowhiz was able to improve the Try facility, which had similar limitations.  Perhaps he can suggest a simple method?

**Comment by Patowhiz** [comment]
> @Wycklife this layout looks neat and I look forward to testing it.
> On your solution for the Try with a single element, I see it works on the command, perhaps because the number, or string is put into an object and then used.
> In a similar situation (File > New Data Frame) @Patowhiz was able to improve the Try facility, which had similar limitations. Perhaps he can suggest a simple method?

@rdstern we discussed with @Wycklife in regards to this. I have looked into the problem and explained it in issue #6018. This dialog can have the original expressions or the modified expressions(if need be). Then we can solve the actual problem in a separate PR.

**Comment by Wycklife** [comment]
@rdstern @dannyparsons Now that there is a new issue on the Try control, Should this now be merged? If it's okay?

**Comment by rdstern** [comment]
@Wycklife this looks good.  I have some trivial corrections to 3 of the examples.  If they could be done tomorrow, then we might be able to get it included in the release.
I have numbered the examples 1 to 10, in their order.
6. begins factor.  Change the final 2 to 3 and add another NA to the end of the tooltip.
8. Begins c( .  Add ,15 at the end.  In the tooltip add 0 at the start and another 15 at the end  - so 0,1,3,5,10,11,12,15,15
10. Change 1940 to 1937.  The tooltip stays as before.

**Comment by Wycklife** [comment]
@rdstern I have fixed the changes above.

**Review by rdstern — DISMISSED** [review]
I am approving, because it all now works.  There are 2 tiny cosmetic points that could be included if done immediately, plus one slightly larger point  that could be done, or left as an issue.
1) In the 8th tooltip there isn't yet a comma before the 10.
2) In the key [ , ] the focus is after the comma and should be before.
Now the bigger point and this has been an error, so far on my part.  We are getting close to being able to suggest using R-Instat to support teaching of R.  Now there is a tidyverse style guide which we should follow.  There are 2 points to note here, the first is in the examples.
After a comma it is good to have a space - as in writing English.
Usually have a space round an operator.
So, in the 3rd example change (2,3,4)*1.5E02 to (2, 3, 4) * 1.5E02.  
Do the same comma rule for the other examples.

Also ready for second approval by @lloyddewit 
I will check any of the minor cosmetic changes above if they are done in time.

**Review by rdstern — CHANGES_REQUESTED** [review]
@Wycklife this is good.  Just a space after the first argument in the last one, i.e. before the as.Date.

And a new one that I hope is trivial too.  We save as Enter1, etc.  The good practice guide says keep to lower case, so change to enter1.
Thanks

**Comment by Wycklife** [comment]
@rdstern Done.

**Review by rdstern — APPROVED** [review]
Great.  Ready for @lloyddewit

File: instat/dlgEnter.vb [hunk 1]
```diff
@@ -309,17 +235,37 @@ Public Class dlgEnter
         lstView.Columns.Add("Commands", 450)
 
         'add rows of sample commands
-        lstView.Items.Add(New ListViewItem({"seq(9 )"}))
-        lstView.Items.Item(0).ToolTipText = "seq(9 )" 'todo. sensible tooltip here.
+        lstView.Items.Add(New ListViewItem({"4.5"}))
+        lstView.Items.Item(0).ToolTipText = "A single number repeated for the data frame" 'todo. sensible tooltip here.
+
+        lstView.Items.Add(New ListViewItem({"'Nairobi'"}))
+        lstView.Items.Item(1).ToolTipText = "a single text repeated for the data frame" 'todo. Sensible tooltip here.
```
Comment (dannyparsons) [inline instat/dlgEnter.vb:242]:
```suggestion
        lstView.Items.Item(1).ToolTipText = "A single text repeated for the data frame" 'todo. Sensible tooltip here.

```

File: instat/dlgEnter.vb [hunk 2]
```diff
@@ -309,17 +235,37 @@ Public Class dlgEnter
...
-        lstView.Items.Add(New ListViewItem({"seq(9 )"}))
-        lstView.Items.Item(0).ToolTipText = "seq(9 )" 'todo. sensible tooltip here.
+        lstView.Items.Add(New ListViewItem({"4.5"}))
+        lstView.Items.Item(0).ToolTipText = "A single number repeated for the data frame" 'todo. sensible tooltip here.
+
+        lstView.Items.Add(New ListViewItem({"'Nairobi'"}))
+        lstView.Items.Item(1).ToolTipText = "a single text repeated for the data frame" 'todo. Sensible tooltip here.
+
+        lstView.Items.Add(New ListViewItem({"c(2, 3, 4)*1.5E02"}))
+        lstView.Items.Item(2).ToolTipText = "gives 300, 450, 600 then repeated for the data frame" 'todo. Sensible tooltip here
```
Comment (dannyparsons) [inline instat/dlgEnter.vb:245]:
```suggestion
        lstView.Items.Item(2).ToolTipText = "Gives 300, 450, 600 then repeated for the data frame" 'todo. Sensible tooltip here

```

File: instat/dlgEnter.vb [hunk 3]
```diff
@@ -309,17 +235,37 @@ Public Class dlgEnter
...
 
-        lstView.Items.Add(New ListViewItem({"LETTERS"}))
-        lstView.Items.Item(1).ToolTipText = "LETTERS" 'todo. sensible tooltip here.
+        lstView.Items.Add(New ListViewItem({"factor(c(1, 2, 3))"}))
+        lstView.Items.Item(3).ToolTipText = "values 1 to 3 with the variable made into a factor" 'todo. Sensible tooltip here
```
Comment (dannyparsons) [inline instat/dlgEnter.vb:248]:
```suggestion
        lstView.Items.Item(3).ToolTipText = "Values 1 to 3 with the variable made into a factor" 'todo. Sensible tooltip here

```

File: instat/dlgEnter.vb [hunk 4]
```diff
@@ -235,27 +187,15 @@ Public Class dlgEnter
 
     Private Sub cmdSequenceFunction_Click_1(sender As Object, e As EventArgs) Handles cmdSequenceFunction.Click
         If chkShowEnterArguments.Checked Then
-            ucrReceiverForEnterCalculation.AddToReceiverAtCursorPosition("seq(from= ,to= ,by= )", 11)
+            ucrReceiverForEnterCalculation.AddToReceiverAtCursorPosition("seq(from= ,to= ,by= ,length =  )", 22)
```
Comment (dannyparsons) [inline instat/dlgEnter.vb:190]: Please follow the [tidyverse style guide](https://style.tidyverse.org/syntax.html), specifically 2.2.1 on commas: always put a space after a comma, and 2.2.4: always put spaces around `=`

File: instat/dlgEnter.vb [hunk 5]
```diff
@@ -309,17 +235,37 @@ Public Class dlgEnter
...
+
+        lstView.Items.Add(New ListViewItem({"c(2, 3, 4)*1.5E02"}))
+        lstView.Items.Item(2).ToolTipText = "gives 300, 450, 600 then repeated for the data frame" 'todo. Sensible tooltip here
 
-        lstView.Items.Add(New ListViewItem({"LETTERS"}))
-        lstView.Items.Item(1).ToolTipText = "LETTERS" 'todo. sensible tooltip here.
+        lstView.Items.Add(New ListViewItem({"factor(c(1, 2, 3))"}))
+        lstView.Items.Item(3).ToolTipText = "values 1 to 3 with the variable made into a factor" 'todo. Sensible tooltip here
+
+        lstView.Items.Add(New ListViewItem({" forcats::as_factor(rep(LETTERS[4:1], c(4, 3, 1, 1)))"}))
```
Comment (dannyparsons) [inline instat/dlgEnter.vb:250]:
```suggestion
        lstView.Items.Add(New ListViewItem({"forcats::as_factor(rep(LETTERS[4:1], c(4, 3, 1, 1)))"}))
```
extra space was included at the start

File: instat/dlgEnter.vb [hunk 6]
```diff
@@ -309,17 +235,37 @@ Public Class dlgEnter
...
+
+        lstView.Items.Add(New ListViewItem({"'Nairobi'"}))
+        lstView.Items.Item(1).ToolTipText = "a single text repeated for the data frame" 'todo. Sensible tooltip here.
+
+        lstView.Items.Add(New ListViewItem({"c(2, 3, 4)*1.5E02"}))
+        lstView.Items.Item(2).ToolTipText = "gives 300, 450, 600 then repeated for the data frame" 'todo. Sensible tooltip here
 
-        lstView.Items.Add(New ListViewItem({"LETTERS"}))
-        lstView.Items.Item(1).ToolTipText = "LETTERS" 'todo. sensible tooltip here.
+        lstView.Items.Add(New ListViewItem({"factor(c(1, 2, 3))"}))
```
Comment (dannyparsons) [inline instat/dlgEnter.vb:247]: Shouldn't this be `forcats::as_factor` if we're now using this?

File: instat/dlgEnter.vb [hunk 7]
```diff
@@ -334,4 +280,99 @@ Public Class dlgEnter
...
+    Private Sub cmdFactor_Click(sender As Object, e As EventArgs) Handles cmdFactor.Click
+        ucrReceiverForEnterCalculation.AddToReceiverAtCursorPosition("forcats::as_factor( )", 2)
+    End Sub
+
+    Private Sub cmdMonth_Click(sender As Object, e As EventArgs)
+        ucrReceiverForEnterCalculation.AddToReceiverAtCursorPosition("'month'")
+    End Sub
+
+    Private Sub cmdRnorm_Click(sender As Object, e As EventArgs)
+        ucrReceiverForEnterCalculation.AddToReceiverAtCursorPosition(" rnorm( )", 2)
```
Comment (dannyparsons) [inline instat/dlgEnter.vb:372]:
```suggestion
        ucrReceiverForEnterCalculation.AddToReceiverAtCursorPosition("rnorm( )", 2)
```
extra space

File: instat/dlgEnter.vb [hunk 8]
```diff
@@ -334,4 +280,99 @@ Public Class dlgEnter
...
+    End Sub
+
+    Private Sub cmdRunif_Click(sender As Object, e As EventArgs)
+        ucrReceiverForEnterCalculation.AddToReceiverAtCursorPosition(" runif( )", 2)
```
Comment (dannyparsons) [inline instat/dlgEnter.vb:376]:
```suggestion
        ucrReceiverForEnterCalculation.AddToReceiverAtCursorPosition("runif( )", 2)
```
extra space

File: instat/dlgEnter.vb [hunk 9]
```diff
@@ -51,38 +51,48 @@ Public Class dlgEnter
...
+
+        'Adding tooltips for the buttons
+        ttEnter.SetToolTip(cmdColon, "A subset, e.g. letters[1:4] gives a, b, c, d.")
+        ttEnter.SetToolTip(cmdLetters2, "The letters, e.g. letters[1:4] gives a, b, c, d. Type LETTERS[1:4] for A, B, C, D.")
+        ttEnter.SetToolTip(cmdFactor, "e.g. forcats::as_factor(""B"", ""c"", ""A"") to make the contents into a factor variable with the levels in the order as entered, so here B is the lowest level.")
+        ttEnter.SetToolTip(cmdRepelicationFunction, "Repeat of a sequence, e.g. rep(c(2,3,4), each=2) gives 2, 2, 3, 3, 4, 4.")
+        ttEnter.SetToolTip(cmdMonths, "month.abb[1:4] is ""Jan2"", ""Feb"", ""Mar"", ""Apr"". Type month.name for full month names.")
+        ttEnter.SetToolTip(cmdText, "Define a character variable, e.g. as.character(c(3, 5, ""a"", ""b"")).")
+        ttEnter.SetToolTip(cmdSequenceFunction, "Sequences, given either as seq(1, 5, 2) to give 1,3,5 or as seq(1,5,length=3) to give the same.")
+        ttEnter.SetToolTip(cmdPi, "The number pi = 3.14.")
```
Comment (dannyparsons) [inline instat/dlgEnter.vb:68]: `Pi` is not equal to 3.14! At least use `3.14...` to make it accurate.

File: instat/dlgEnter.vb [hunk 10]
```diff
@@ -309,17 +235,37 @@ Public Class dlgEnter
         lstView.Columns.Add("Commands", 450)
 
         'add rows of sample commands
-        lstView.Items.Add(New ListViewItem({"seq(9 )"}))
-        lstView.Items.Item(0).ToolTipText = "seq(9 )" 'todo. sensible tooltip here.
+        lstView.Items.Add(New ListViewItem({"4.5"}))
+        lstView.Items.Item(0).ToolTipText = "A single number repeated for the data frame" 'todo. sensible tooltip here.
+
+        lstView.Items.Add(New ListViewItem({"'Nairobi'"}))
```
Comment (dannyparsons) [inline instat/dlgEnter.vb:241]: Here you've got single quotes whereas in other places you have double quotes. Is that deliberate?

**Review by dannyparsons — CHANGES_REQUESTED** [review]
Please check for consistency in the R code (following the tidyverse style guide https://style.tidyverse.org/syntax.html). I commented on some below but please check throughout.
Also, please check on grammar and consistency in tooltips. I think tooltips should be sentences so start with a capital letter and end with a full stop.

File: instat/dlgEnter.vb (hunk 10 above)
Comment (Wycklife) [reply instat/dlgEnter.vb:241]: @dannyparsons This what Roger suggested. Do you want it changed?

**Comment by Wycklife** [comment]
@dannyparsons Could you check if your changes are properly fixed? Thank you

**Comment by rdstern** [comment]
I am always confused on the use of single and double quotes.  But we should try to be consistent one way or the other!

**Comment by Wycklife** [comment]
> I am always confused on the use of single and double quotes. But we should try to be consistent one way or the other!

Meaning I should I should change items with single quotes to have double quotes?

**Comment by rdstern** [comment]
Meaning that on items of programming and consistency, if @dannyparsons and I differ, then usually follow danny!  Does that make it clear?  If so, then let me know the decision!

**Comment by lloyddewit** [comment]
@Wycklife There are a number of open comments above. Please could you respond to each comment stating which action you took. If you changed code in response to the comment then please state what you changed (preferably with line number). Currently I have to do detective work to link the updated code to the comment to the supdated code. This is time-consuming and error-prone. 
If you do the above then I can do the final review, thanks

File: instat/dlgEnter.vb (hunk 2 above)
Comment (Wycklife) [reply instat/dlgEnter.vb]: @dannyparsons This has been fixed

File: instat/dlgEnter.vb (hunk 4 above)
Comment (Wycklife) [reply instat/dlgEnter.vb]: @dannyparsons I have done the spacing

File: instat/dlgEnter.vb (hunk 7 above)
Comment (Wycklife) [reply instat/dlgEnter.vb]: @dannyparsons Spaces removed.

File: instat/dlgEnter.vb (hunk 9 above)
Comment (Wycklife) [reply instat/dlgEnter.vb]: @dannyparsons I have changed this.

File: instat/dlgEnter.vb (hunk 10 above)
Comment (Wycklife) [reply instat/dlgEnter.vb]: @dannyparsons I have added double quotes.

File: instat/dlgEnter.vb (hunk 5 above)
Comment (Wycklife) [reply instat/dlgEnter.vb]: @dannyparsons This extra space has been removed

File: instat/dlgEnter.vb (hunk 8 above)
Comment (Wycklife) [reply instat/dlgEnter.vb]: @dannyparsons The extra space has been removed.

File: instat/dlgEnter.vb (hunk 3 above)
Comment (Wycklife) [reply instat/dlgEnter.vb]: @dannyparsons I have resolved this

File: instat/dlgEnter.vb (hunk 6 above)
Comment (Wycklife) [reply instat/dlgEnter.vb:247]: @rdstern Should I add forcats::as_factor instead of factor for consistency?

**Comment by Wycklife** [comment]
@lloyddewit I have resolved the comments, please could it now be merged?

File: instat/dlgEnter.vb (hunk 6 above)
Comment (lloyddewit) [reply instat/dlgEnter.vb:247]: @Wycklife I don't see a response from @rdstern. Was this comment resolved in an external conversation (if so, please could you document here)? Or is this comment still open?

**Comment by lloyddewit** [comment]
@Wycklife Thank you for clarifying the resolved comments. There is still one comment above that may still be unresolved, please can you clarify? thanks

File: instat/dlgEnter.vb (hunk 6 above)
Comment (Wycklife) [reply instat/dlgEnter.vb:247]: @lloyddewit I have discussed this with Roger and he has agreed that the change should be made. I have done that and pushed the minor change.

**Comment by Wycklife** [comment]
@lloyddewit I have made that small change. The change is about replacing factor() with forcats::as_factor() in line 253. Thank you. The comments have been resolved.

**Review by lloyddewit — APPROVED** [review]

**Comment by lloyddewit** [comment]
@Wycklife Thanks for the final change, looks good.

@rdstern there are 2 changes since your last approval (replaced single quotes with double quotes and replaced `factor()` with `forcats::as_factor()`. Please could you approve? thanks

**Review by rdstern — APPROVED** [review]
Good.  Approved.

**Pull request merged**